pkg/controller/machineset.go

```diff
@@ -677,6 +694,75 @@ func (c *controller) terminateMachines(inactiveMachines []*v1alpha1.Machine, mac
 	return nil
 }
 
+func (c *controller) terminateStaleMachines(inactiveMachines []*v1alpha1.Machine, deletionWindow int, machineSet *v1alpha1.MachineSet) error {
+	var (
+		wg    sync.WaitGroup
+		errCh = make(chan error, deletionWindow)
+	)
+	defer close(errCh)
+
+	wg.Add(deletionWindow)
+	for i := 0; i < deletionWindow; i++ {
+		go c.prepareMachineForDeletion(inactiveMachines[i], machineSet, &wg, errCh)
+	}
+	wg.Wait()
+
+	select {
+	case err := <-errCh:
+		// all errors have been reported before and they're likely to be the same, so we'll only return the first one we hit.
+		if err != nil {
+			return err
+		}
+	default:
+	}
+
+	return nil
+}
+
+// calculateStaleMachineDeletionWindow computes
+func (c *controller) calculateStaleMachineDeletionWindow(activeMachines, runningMachines, inactiveMachines []*v1alpha1.Machine, machineSet *v1alpha1.MachineSet) int {
+	var (
+		deletionWindow          int
+		numOfActiveMachines     = len(activeMachines)
+		numOfRunningMachines    = len(runningMachines)
+		numOfInactiveMachines   = len(inactiveMachines)
+		numOfMachineSetReplicas = int(machineSet.Spec.Replicas)
+	)
+
+	// nothing to do
+	if numOfInactiveMachines == 0 {
+		return 0
+	}
+
+	// Will not delete any stale machines. It is expected that the user deletes these machines manually.
+	if c.safetyOptions.FailedMachineDeletionRatio == 0 {
+		return 0
+	}
+
+	if c.safetyOptions.FailedMachineDeletionRatio == 1 {
+		return numOfInactiveMachines
+	}
+
+	if numOfMachineSetReplicas == 0 {
+		deletionWindow = numOfInactiveMachines
+	} else {
+		deletionWindow = int(math.Round(float64(numOfMachineSetReplicas) * c.safetyOptions.FailedMachineDeletionRatio))
+		if deletionWindow == 0 {
+			deletionWindow = 1
+		}
+
+		pendingMachines := numOfActiveMachines - numOfRunningMachines
+
+		if deletionWindow > pendingMachines {
```

Comment:
@zuzzas Can you please explain the reasoning behind comparing with `pendingMachines`, I somehow failed to understand it.

Reply:
I am with @hardikdr on this. I am failing to understand the logic here. You could perhaps, help us why you are not taking the calculations with just `inActiveMachines (Failed)` machines? Is there a case you are trying to cover that we have missed. Explanation with an example would be helpful.

Reply:
`pendingMachines` is a horrible variable name, since it throws everyone off due to collision with the `Pending` Machines state. I've renamed it into a longer, but a proper name `numOfActiveNotRunningMachines`.

`Pending`, `Unknown`, `Terminating` are transient Machine states. Machines transition through them into more concrete `Running` or `Failed` states. I believe that we should consider all Machines in these transient states when we calculate `deletionWindow`. This way we won't delete Failed Machines until Machines in the aforementioned transient states stabilize into `Running` or `Failed` state.

Btw, `Available` state is not used anywhere in the codebase (might be a candidate for deletion?).

Reply:
Resolving this, as code is changed now.